### PR TITLE
Creating idminter service

### PR DESCRIPTION
**Pull request by alicefuzier**
Creating a service that generates IDs for resources and checks for overlaps in DynamoDB. 
Addresses issue #123

### CODE REVIEW

File: ingestor/src/main/scala/uk/ac/wellcome/ingestor/modules/SQSWorker.scala [hunk 1]
```diff
@@ -1,60 +1,25 @@
...
-}
+import uk.ac.wellcome.sqs.SQSReader
+import uk.ac.wellcome.utils.TryBackoff
+import uk.ac.wellcome.utils.GlobalExecutionContext.context
 
 object SQSWorker extends TwitterModule with TryBackoff {
-  override val modules = Seq(SQSConfigModule, SQSClientModule, AkkaModule)
 
   val waitTime = flag("sqs.waitTime", 20, "SQS wait time")
   val maxMessages = flag("sqs.maxMessages", 1, "Max SQS messages")
```
Comment (alicefuzier) [inline ingestor/src/main/scala/uk/ac/wellcome/ingestor/modules/SQSWorker.scala]: delete these?

File: id_minter/conf/application.dev.conf [hunk 2]
```diff
@@ -0,0 +1,4 @@
+es {
```
Comment (alicefuzier) [inline id_minter/conf/application.dev.conf]: delete this file?Or create it empty

File: id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/modules/IdMinterModule.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import uk.ac.wellcome.sns.SNSWriter
+import uk.ac.wellcome.sqs.SQSReader
+import uk.ac.wellcome.utils.{JsonUtil, TryBackoff}
+import uk.ac.wellcome.utils.GlobalExecutionContext.context
+
+
+object IdMinterModule extends TwitterModule with TryBackoff{
+
+  override def singletonStartup(injector: Injector) {
+    info("Starting SQS worker")
```
Comment (alicefuzier) [inline id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/modules/IdMinterModule.scala]: this log is wrong

File: common/src/main/scala/uk/ac/wellcome/models/UnifiedItem.scala [hunk 4]
```diff
@@ -3,12 +3,16 @@ package uk.ac.wellcome.models
 import com.sksamuel.elastic4s._
 import uk.ac.wellcome.utils.JsonUtil
 
-case class Identifier(source: String, sourceId: String, value: String)
+case class Identifier(CanonicalID: String, MiroID: String)
+
+case class IdentifiedUnifiedItem(canonicalId: String, unifiedItem: UnifiedItem)
+
+case class SourceIdentifier(source: String, sourceId: String, value: String)
```
Comment (alexwlchan) [inline common/src/main/scala/uk/ac/wellcome/models/UnifiedItem.scala]: Perhaps add a couple of Scaladoc comments for additional clarity? Something like:

```scala
/** A set of identifiers as stored in DynamoDB */
case class Identifier

/** An identifier received from one of the original sources */
case class SourceIdentifier

/** A representation of an item in our ontology, without a canonical identifier */
case class UnifiedItem
```

Just a suggestion.

File: common/src/test/scala/uk/ac/wellcome/test/utils/SNSLocal.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      :topic_arn: arn:aws:sns:us-east-1:123456789012:es_ingest
+      :structure:
+      :target_arn:
+      :received_at: 2017-04-10 16:31:38.015426145 +00:00
+
+     */
+
+    val string = scala.io.Source.fromURL(localSNSEndpointUrl).mkString
+    string
+      .substring(string.indexOf("- :id:") + 1)
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/test/utils/SNSLocal.scala]: I’m not super familiar with Scala’s string methods – can we talk about how this works?

File: common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      }
+      messages.head should not be equal(messages.tail.head)
+    }
+  }
+
+  it("should return a failed future if writing to the SNS topic fails") {
+    val sqsConfig =
+      SQSConfig("eu-west-1", "not a valid queue url", waitTime = 20 seconds, 1)
+    val sqsReader =
+      new SQSReader(sqsClient, sqsConfig)
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala]: nit: pull onto a single line?

File: common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      SQSConfig("eu-west-1", idMinterQueueUrl, waitTime = 20 seconds, 2)
+    val messageStrings = List("someMessage1", "someMessage2", "someMessage3")
+    messageStrings.foreach(sqsClient.sendMessage(idMinterQueueUrl, _))
+    val sqsReader =
+      new SQSReader(sqsClient, sqsConfig)
+
+    val futureMessages = sqsReader.retrieveMessages()
+
+    whenReady(futureMessages) { messages =>
+      messages should have size 2
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala]: Maybe worth a quick comment here about why we aren’t doing the more obvious test?

> SQS isn’t a FIFO queue, so we can’t just check we get "someMessage1" and "someMessage2". Instead we have to test that we receive two distinct messages, both of which were in the original set of messages.

File: common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    with Matchers
+    with ScalaFutures
+    with IntegrationPatience
+    with SQSLocal {
+
+  it("should get messages from the SQS queue, limited by the maximum number of messages and return them") {
+    val sqsConfig =
+      SQSConfig("eu-west-1", idMinterQueueUrl, waitTime = 20 seconds, 2)
+    val messageStrings = List("someMessage1", "someMessage2", "someMessage3")
+    messageStrings.foreach(sqsClient.sendMessage(idMinterQueueUrl, _))
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala]: Nice change. 👍

File: common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+class SQSReaderTest
+    extends FunSpec
+    with Matchers
+    with ScalaFutures
+    with IntegrationPatience
+    with SQSLocal {
+
+  it("should get messages from the SQS queue, limited by the maximum number of messages and return them") {
+    val sqsConfig =
+      SQSConfig("eu-west-1", idMinterQueueUrl, waitTime = 20 seconds, 2)
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala]: Can we add a `maxMessages =` to this SQSConfig constructor please? Improves clarity and readability.

File: common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      messages.foreach { message =>
+        messageStrings should contain(message.getBody)
+      }
+      messages.head should not be equal(messages.tail.head)
+    }
+  }
+
+  it("should return a failed future if writing to the SNS topic fails") {
+    val sqsConfig =
+      SQSConfig("eu-west-1", "not a valid queue url", waitTime = 20 seconds, 1)
```
Comment (alexwlchan) [inline common/src/test/scala/uk/ac/wellcome/sqs/SQSReaderTest.scala]: ditto above

**Pull request merged**